### ISSUE TITLE
Tracking results differ depending on batch size despite persist

### ISSUE THREAD

**Issue by antoine-metaphysic** [ultralytics/ultralytics#19288]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.


### Ultralytics YOLO Component

Predict

### Bug

Hello,

To reduce memory constraints, I am trying to call `model.track(..., persist=True)` with batches of frames rather than a single call with all the frames of a video. But while testing, I've noticed that the tracking results differ when using the batching call.

See example below. The results when passing batches of 15 frames (`results1`) is visibly worse than when passing all frames at once (`results2`), with detection dropping out every 15 frames, and the tracks getting sometimes split.

`results2` (single batch of 120 frames), the person with id 1009 gets consistently tracked throughout:

https://github.com/user-attachments/assets/cf095968-7e7d-473b-8cac-42aec0888d04

`results1` (multiple batches of 15 frames), the same person gets 3 different ids throughout the video, the bounding box is jumping around and sometimes dropped entirely:

https://github.com/user-attachments/assets/28a43df5-872e-435f-b47c-98a0279e1fc2

My understanding from the [documentation](https://docs.ultralytics.com/modes/track/#persisting-tracks-loop) was that by using `persist=True` I could split inference into multiple calls while still getting the same results, am I reading this wrong? Is there no way to get the same quality while splitting processing in batches?

Thanks,
A.

### Environment

```
Ultralytics 8.3.75 🚀 Python-3.12.0 torch-2.6.0+cu124 CUDA:0 (NVIDIA A100-SXM4-40GB, 40507MiB)                                                                                                                       │   │                                                                                                                                                                                                       MEM 1│
Setup complete ✅ (24 CPUs, 176.9 GB RAM, 772.1/983.7 GB disk)                                                                                                                                                       │   │                                                                                                                                                                                                            │
                                                                                                                                                                                                                     │75%│                                                                                                                                                                                                            │
OS                  Linux-5.15.0-131-generic-x86_64-with-glibc2.31                                                                                                                                                   │   │                                                                                                                                                                                                            │
Environment         Linux                                                                                                                                                                                            │   │                                                                                                                                                                                                            │
Python              3.12.0                                                                                                                                                                                           │   │                                                                                                                                                                                                            │
Install             pip                                                                                                                                                                                              │   │                                                                                                                                                                                                            │
RAM                 176.88 GB                                                                                                                                                                                        │50%│                                                                                                                                                                                                            │
Disk                772.1/983.7 GB                                                                                                                                                                                   │   │                                                                                                                                                                                                            │
CPU                 Intel Xeon Gold 6342 2.80GHz                                                                                                                                                                     │   │                                                                                                                                                                                                            │
CPU count           24                                                                                                                                                                                               │   │                                                                                                                                                                                                            │
GPU                 NVIDIA A100-SXM4-40GB, 40507MiB                                                                                                                                                                  │   │                                                                                                                                                                                                            │
GPU count           2                                                                                                                                                                                                │25%│                                                                                                                                                                                                            │
CUDA                12.4                                                                                                                                                                                             │   │                                                                                                                                                                                                            │
                                                                                                                                                                                                                     │   │                                                                                                                                                                                                            │
numpy               ✅ 2.1.1<=2.1.1,>=1.23.0                                                                                                                                                                         │   │────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────│
matplotlib          ✅ 3.10.0>=3.3.0                                                                                                                                                                                 │   │                                                                                                                                                                                                            │
opencv-python       ✅ 4.11.0.86>=4.6.0                                                                                                                                                                              │ 0%│────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────│
pillow              ✅ 11.1.0>=7.1.2                                                                                                                                                                                 │   └────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
pyyaml              ✅ 6.0.2>=5.3.1                                                                                                                                                                                  │  PID       USER GPU    TYPE       GPU MEM    CPU  HOST MEM Command                                                                                                                                                 
requests            ✅ 2.32.3>=2.23.0                                                                                                                                                                                │ 2093 paperspace   0 Graphic     89MB   0%     0%     357MB /usr/bin/gnome-shell                                                                                                                                    
scipy               ✅ 1.15.2>=1.4.1                                                                                                                                                                                 │ 1872       root   0 Graphic     72MB   0%     0%      91MB /usr/lib/xorg/Xorg vt2 -displayfd 3 -auth /run/user/1000/gdm/Xauthority -background none -noreset -keeptty -verbose 3
torch               ✅ 2.6.0>=1.8.0                                                                                                                                                                                  │ 1872       root   1 Graphic      4MB   0%     0%      91MB /usr/lib/xorg/Xorg vt2 -displayfd 3 -auth /run/user/1000/gdm/Xauthority -background none -noreset -keeptty -verbose 3
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"                                                                                                                                                 │43054 paperspace   0 Compute   2760MB   7%     0%    1996MB python -m ipykernel_launcher -f /home/paperspace/.local/share/jupyter/runtime/kernel-7411ba0f-b09d-4f03-997c-ff50f31cc279.json
torchvision         ✅ 0.21.0>=0.9.0                                                                                                                                                                                 │
tqdm                ✅ 4.67.1>=4.64.0                                                                                                                                                                                │
psutil              ✅ 7.0.0                                                                                                                                                                                         │
py-cpuinfo          ✅ 9.0.0                                                                                                                                                                                         │
pandas              ✅ 2.2.3>=1.1.4                                                                                                                                                                                  │
seaborn             ✅ 0.13.2>=0.11.0                                                                                                                                                                                │
ultralytics-thop    ✅ 2.0.14>=2.0.0   
```

### Minimal Reproducible Example

```python
import os
import itertools
import tempfile
import numpy as np
import torch
import torchvision
from torchvision.datasets.utils import download_url

download_url("https://github.com/pytorch/vision/blob/main/test/assets/videos/SOX5yA1l24A.mp4?raw=true", ".", "SOX5yA1l24A.mp4")
reader = torchvision.io.VideoReader("./SOX5yA1l24A.mp4", "video")
fps = float(reader.get_metadata()["video"]["fps"][0])
reader.set_current_stream("video")
frames = [frame['data'] for frame in reader]
video = torch.stack(frames)[0:120, ..., :320]/255.

from ultralytics import YOLO
model = YOLO("yolo11n.pt")

def track(video, split):
    return list(itertools.chain(*(model.track(chunk, persist=True, verbose=False, stream=True, stream_buffer=True, tracker="bytetrack.yaml") for chunk in video.split(split))))

results1 = track(video, 15)

results2 = track(video, 120)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @antoine-metaphysic, thank you for sharing such a detailed report and for your interest in Ultralytics 🚀!

We appreciate your efforts to reproduce the issue and provide detailed information. While our engineering team reviews this, here's how you can assist further:

1. For bug reports 🐛, please ensure your reproduction code is a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help pinpoint the root cause.
2. Based on your description of discrepancies in the tracking behavior, could you confirm if:
   - The `tracker_config` is consistent when processing in batches vs. as a whole?
   - There are dependency mismatches or differences in the versions (e.g., PyTorch, torchvision) across systems that could affect GPU processing?

We encourage you to upgrade to the latest version of the `ultralytics` package and dependencies to ensure you're leveraging all the latest fixes and improvements. Run the following command:

```bash
pip install -U ultralytics
```

For further debugging, check if this issue is present in our verified deployment environments, such as Google Colab, Kaggle, Paperspace, or Docker. Here are links for easy access and quick execution:

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Docker: <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For general questions about tracking or training, we also suggest checking our [documentation on YOLO tracking](https://docs.ultralytics.com/modes/track/) and ensuring you are following our [best practices for model training](https://docs.ultralytics.com/guides/model-training-tips/). This might help clarify expected behaviors such as `persist=True`.

Join the Ultralytics community for real-time discussion and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://reddit.com/r/Ultralytics)

Finally, if you have access to additional sample videos or logs, feel free to share them to aid troubleshooting. Rest assured, an Ultralytics engineer will follow up soon to assist you further! 😊

**Comment by Y-T-G** [comment]
Each image in the batch is considered part of a separate stream, so each element is being tracked with  a separate tracker.

First element will use tracker1, second element tracker2 and so on. That's the effect of passing batched images to `model.track()`.

**Comment by Y-T-G** [comment]
You should reinitialize the model after each test.

**Comment by Y-T-G** [comment]
I can see a difference with batching. But it seems to be an issue with batch prediction, and not tracking.

**Comment by Y-T-G** [comment]
Not sure if it's bug, or it's a floating point error.

```python
In [5]: import torch
In [6]: model = YOLO("yolov8n.pt").cuda()
In [7]: torch.allclose(model.model(torch.zeros(2, 3, 640, 640).cuda())[0][:1], model.model(torch.zeros(16, 3, 640, 640).cuda())[0][:1], 1e-2)
Out[7]: False
```

**Comment by Y-T-G** [comment]
Doesn't seem like a bug. The output of `Conv2D` varies based on different batch size.

https://stackoverflow.com/questions/76864410/why-conv2d-yields-different-results-with-different-batch-size

**Comment by Y-T-G** [comment]
To verify, it's not due to tracker, I replaced the batch 16 inference results with the results from batch 1 inference, while still having the tracker run on batches:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

results = model("ped.mp4", stream=True, batch=1, save=False, verbose=False)

preds = []
for result in results:
    result.boxes.data.cpu()
    preds.append(result)

i = 0
def postprocess(self, *args, **kwargs):
    results = []
    global i
    for b in range(1):
        result = preds[i + b]
        result.boxes.data.cuda()
        results.append(result)
    i += 1
    return results

model.predictor.postprocess = postprocess.__get__(model.predictor, model.predictor.__class__)
results = model.track("ped.mp4", verbose=False, persist=True, tracker="bytetrack.yaml")

results_b1 = [r.boxes.id.tolist() for r in results]

del model.predictor.trackers
model.callbacks["on_predict_start_end"].clear()
model.callbacks["on_predict_postprocess_end"].clear()

i = 0
def postprocess(self, *args, **kwargs):
    results = []
    global i
    for b in range(16):
        try:
            result = preds[i + b]
        except IndexError:
            return results
        result.boxes.data.cuda()
        results.append(result)
    i += 16
    return results

model.predictor.postprocess = postprocess.__get__(model.predictor, model.predictor.__class__)
results = model.track("ped.mp4", verbose=False, persist=True, batch=16, tracker="bytetrack.yaml")

results_b16 = [r.boxes.id.tolist() for r in results]

print(all([r1 == r2 for r1, r2 in zip(results_b1, results_b16)]))
```

And the output is the same.

**Comment by antoine-metaphysic** [comment]
My impression is that the detector is impacted by the batch size: it's quite obvious in both cases that the first couple frames only contain a subset of all the objects, and it only detects most objects from the third frame onwards. So if the detector doesn't rely on `persist`, then it will miss detection at the start of every batch, and that will in turn impact the tracker, as there will be missing bounding boxes. That would also explain why even when it manages to link the tracklets correctly, the tracked boxes in the batched case are visibly more unstable than in the non-batched case.

**Comment by Y-T-G** [comment]
> My impression is that the detector is impacted by the batch size: it's quite obvious in both cases that the first couple frames only contain a subset of all the objects, and it only detects most objects from the third frame onwards.

Any object that hasn't been assigned a track ID wouldn't be returned even if they were detected unless it's the first frame or unless none of the objects received track ID.

I'm alao not sure about the video. The first question is why does it start from 700 or 1000+ ID? If those are the first frames, the ID should have started from 0 or close to that.

The second problem is, you should be resetting the tracker after the fies test, which is not being done in the MRE.

**Comment by antoine-metaphysic** [comment]
> I'm also not sure about the video. The first question is why does it start from 700 or 1000+ ID? If those are the first frames, the ID should have started from 0 or close to that.
> 
> The second problem is, you should be resetting the tracker after the first test, which is not being done in the MRE. It's carrying the state from the previous test.

Both issues came from not resetting the tracker as you recommended. Below are results recreating the model each time, the IDs now start from 1 and the behaviour is otherwise the same:

One big batch:
https://github.com/user-attachments/assets/bd514156-39dd-4975-b53e-57b8576a8db8

Multiple smaller batches:
https://github.com/user-attachments/assets/51abb674-1c0b-4fdf-98cb-922affd180d1

Full example in colab: https://colab.research.google.com/drive/1ielj1y2PV7PwOiUpuJHLzs9xMYUoF8aW?usp=sharing

**Comment by Y-T-G** [comment]
Can you label the frame number on the video? And also vertically stack them into the same video.

**Comment by antoine-metaphysic** [comment]
Sure thing! Here you go:

https://github.com/user-attachments/assets/f12a666f-8b03-465b-8c49-5e2e7c9abd05

I also ran the same test on just the `predict` call, and this one shows no difference between the two cases:

https://github.com/user-attachments/assets/ea7e4799-9af3-479d-8afb-6b3cac502032

So prediction differences aren't the cause of the tracking differences after all.

With the frame number labels, the differences in the tracking results can be clearly seen at the start of almost each batch.

The results come from the updated colab notebook, feel free to play with it: https://colab.research.google.com/drive/1ielj1y2PV7PwOiUpuJHLzs9xMYUoF8aW?usp=sharing

**Comment by Y-T-G** [comment]
The behavior seems to be caused by the particular way you are generating the batched tracking results. It seems to be caused by `itertools.chain()` creating a "snapshot" of the model object such that each batch runs with a model object that doesn't carry the state changes from the previous runs.

This results in several issues, including duplicated callbacks which get added based on whether `model.predictor.trackers` exist (which ought to exist after the first time `model.track()` is called):
```python
model.predictor.callbacks["on_predict_postprocess_end"]
```
```
[<function ultralytics.utils.callbacks.base.on_predict_postprocess_end(predictor)>,
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True),
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True)]
```
It doesn't occur if the tracking is done without `itertools.chain()`:
```python
def track(video, split):
    model = YOLO("yolo11n.pt")
    results = [list(model.track(chunk, persist=True, verbose=False, stream=True, stream_buffer=True, tracker="bytetrack.yaml")) for chunk in video.split(split)]
    return model, np.stack([r.plot() for result in results for r in result])
```
```
[<function ultralytics.utils.callbacks.base.on_predict_postprocess_end(predictor)>,
 functools.partial(<function on_predict_postprocess_end at 0x7f514cbbb9c0>, persist=True)]
```

So each batch in your code is creating a new tracker, restarting the tracking session.

**Comment by antoine-metaphysic** [comment]
Oh, that's fascinating, thanks for the insight. Definitely not something I was expecting.

It looks like even without `itertools.chain`, it doesn't take much to get the wrong behaviour. E.g., this also results in the tracker restarting:

```
results = list(model.track(chunk, persist=True, verbose=False, stream=True, stream_buffer=True, tracker="bytetrack.yaml") for chunk in video.split(split))
```

I would think that's a pretty innocuous use of the model, and hard to notice something might go wrong. Maybe there's a way to make this interface more robust?

In the meantime, I now have a workaround, thanks a lot for this.

**Comment by antoine-metaphysic** [comment]
To be clear, the workaround here is not whether or not `itertools.chain()` is used, but the fact that the `model.track()` call (which returns a generator) is immediately evaluated (thanks to the `list(model.track(...))` wrapper) before the next call. So the following also works:

```
args = dict(persist=True, verbose=False, stream=True, stream_buffer=True, tracker="bytetrack.yaml")
results = itertools.chain(*(list(model.track(chunk, **args)) for chunk in video.split(split)))
return np.stack([r.plot() for r in results])
```

Alternatively, if I take out the `stream=True` argument (which I realize I don't need in this case), the `model.track()` call doesn't return a generator and therefore the original call works fine:

```
args = dict(persist=True, verbose=False, stream_buffer=True, tracker="bytetrack.yaml")
results = itertools.chain(*(model.track(chunk, **args) for chunk in video.split(split)))
return np.stack([r.plot() for r in results])
```

So I guess the crux is the results of `model.track(..., stream=True)` should all be consumed before making another call to it.

**Comment by glenn-jocher** [comment]
Correct, the immediate evaluation of `model.track(..., stream=True)` before subsequent calls is crucial for maintaining the tracker's state when processing video in chunks.

**Comment by antoine-metaphysic** [comment]
Thanks for helping clear that up and troubleshoot my issue.
